File: ui/src/models/mcp-tool.ts
import { type BasePaginatorResponse, type BaseResponse } from '@/models/base'

export type McpToolProvider = {
  id: string
  name: string
  description: string
  transport: string
  url: string
  headers: Record<string, string>
  tools: Array<any>
  updated_at: number
  created_at: number
}

export type GetMcpToolProvidersWithPageResponse = BasePaginatorResponse<McpToolProvider>

export type CreateMcpToolProviderRequest = {
  mcp_schema: string
}

export type UpdateMcpToolProviderRequest = {
  mcp_schema: string
}

export type GetMcpToolProviderResponse = BaseResponse<
  McpToolProvider & {
    config: Record<string, any>
    mcp_schema: string
  }
>

export type GetMcpToolResponse = BaseResponse<{
  id: string
  name: string
  description: string
  input_schema: Record<string, any>
  metadata: Record<string, any>
  provider: {
    id: string
    name: string
    label: string
    icon: string
    description: string
    transport: string
    url: string
    headers: Record<string, string>
  }
  inputs: {
    type: string
    name: string
    required: boolean
    description: string
    schema: Record<string, any>
  }[]
}>

File: api/internal/model/mcp_tool.py
#!/user/bin/env python
# -*- coding: utf-8 -*-
"""
@Time   : 2026/6/22 
@Author : wzy
@File   : mcp_tool
"""
from datetime import datetime

from sqlalchemy import (
    Column,
    UUID,
    String,
    Text,
    DateTime,
    text,
    PrimaryKeyConstraint,
    Index,
)
from sqlalchemy.dialects.postgresql import JSONB

from internal.extension.database_extension import db


class McpToolProvider(db.Model):
    """MCP工具提供者模型"""
    __tablename__ = "mcp_tool_provider"
    __table_args__ = (
        PrimaryKeyConstraint("id", name="pk_mcp_tool_provider_id"),
        Index("mcp_tool_provider_account_id_idx", "account_id"),
        Index("mcp_tool_provider_name_idx", "name"),
    )

    id = Column(UUID, nullable=False, server_default=text("uuid_generate_v4()"))
    account_id = Column(UUID, nullable=False)
    name = Column(String(255), nullable=False, server_default=text("''::character varying"))
    description = Column(Text, nullable=False, server_default=text("''::text"))
    transport = Column(String(255), nullable=False, server_default=text("'http'::character varying"))
    url = Column(String(1024), nullable=False, server_default=text("''::character varying"))
    headers = Column(JSONB, nullable=False, server_default=text("'{}'::jsonb"))
    config = Column(JSONB, nullable=False, server_default=text("'{}'::jsonb"))
    mcp_schema = Column(Text, nullable=False, server_default=text("''::text"))
    updated_at = Column(
        DateTime,
        nullable=False,
        server_default=text("CURRENT_TIMESTAMP(0)"),
        onupdate=datetime.now,
    )
    created_at = Column(DateTime, nullable=False, server_default=text("CURRENT_TIMESTAMP(0)"))

    @property
    def tools(self) -> list["McpTool"]:
        return db.session.query(McpTool).filter_by(provider_id=self.id).all()


class McpTool(db.Model):
    """MCP工具表"""
    __tablename__ = "mcp_tool"
    __table_args__ = (
        PrimaryKeyConstraint("id", name="pk_mcp_tool_id"),
        Index("mcp_tool_account_id_idx", "account_id"),
        Index("mcp_tool_provider_id_name_idx", "provider_id", "name"),
    )

    id = Column(UUID, nullable=False, server_default=text("uuid_generate_v4()"))
    account_id = Column(UUID, nullable=False)
    provider_id = Column(UUID, nullable=False)
    name = Column(String(255), nullable=False, server_default=text("''::character varying"))
    description = Column(Text, nullable=False, server_default=text("''::text"))
    input_schema = Column(JSONB, nullable=False, server_default=text("'{}'::jsonb"))
    tool_metadata = Column(JSONB, nullable=False, server_default=text("'{}'::jsonb"))
    updated_at = Column(
        DateTime,
        nullable=False,
        server_default=text("CURRENT_TIMESTAMP(0)"),
        onupdate=datetime.now,
    )
    created_at = Column(DateTime, nullable=False, server_default=text("CURRENT_TIMESTAMP(0)"))

    @property
    def provider(self) -> McpToolProvider:
        """只读属性，返回当前工具关联/归属的MCP工具提供者信息"""
        return db.session.get(McpToolProvider, self.provider_id)
